Adding new files and Changes for the JSON format feature

BitsonFire opened the pull request:
This PR is with regards to fix #2 issue.

I've added in the files and have made changes as required to include the json format output feature. JSON format can be enabled by using -j in the command line parameters and if not specified, text format will be shown by default. Please let me know if it is correct or if any changes are required.

Thank You!

gavv commented:
Thanks for the PR! I'll review & test it in upcoming days.

gavv commented on src/LatencyEstimator.cpp:
```diff
@@ -37,6 +38,11 @@ LatencyEstimator::LatencyEstimator(const Config& config)
     , output_trigger_(config_)
     , input_trigger_(config_)
     , sma_(config.sma_window) {
+        config_.enable_json ? format = new JSONFormatter : format = new TextFormatter;
```
Instead of duplicating this check in LatencyEstimator and LossEstimator, I suggest to do it once in main() and then just pass IFormatter to estimator constructor. Since formatter should be always present, it may be a reference, not a pointer.

gavv commented on src/Config.hpp:
```diff
@@ -16,6 +16,9 @@ struct Config {
     // 16-bit little-endian signed integers
     snd_pcm_format_t format { SND_PCM_FORMAT_S16_LE };
 
+    // Enable JSON formatted output
+    bool enable_json { false };
```
If future we may want to add other output formats. So instead of a boolean, I suggest to add a enum, something like Formatting, FormatterType, OutputFormat, etc. with two values (for text and json).

Also, a nitpick: the fields access/format/sample_rate/channels are related and thus grouped together, so maybe better move the new filed to the beginning or to the end of the struct.

gavv commented on src/JSONFormatter.hpp:
```diff
@@ -0,0 +1,20 @@
+
+        void report_losses(double loss_rate, int sma_window, double avg_loss_rate, double loss_ratio) override;
+
+        bool first_output;
```
This should be private (and thus renamed to `first_output_`).

gavv commented on src/JSONFormatter.cpp:
```diff
@@ -0,0 +1,35 @@
+        first_output = false;
+    }
+
+    se_log_info("  {\"sw+hw\": %7.3f, \"hw\": %7.3f, \"hw_avg%d\": %7.3f},", sw_hw,
+        hw, sma_window, hw_avg);
```
I think we should switch from se_log_info to printf in both formatters. se_log_info prints to stderr, but ideally we should print only logs to stderr and the measurement results should be printed to stdout, so that they can be easily redirected to file.

gavv commented on src/JSONFormatter.cpp:
```diff
@@ -0,0 +1,35 @@
+        first_output = false;
+    }
+
+    se_log_info("  {\"rate\": %5.1f, \"rate_avg%d\": %5.1f, \"ratio\": %6.2f%%},",
+            loss_rate, sma_window, avg_loss_rate, loss_ratio);
```
JSON does not allow trailing comma in arrays. This is incorrect:

```
[
  "foo",
  "bar",
]
```

This is correct:

```
[
  "foo",
  "bar"
]
```

So we should print the line without comma and \n and flush it. Then, if a new line is printed, we should print ",\n", and otherwise we should print "\n]\n".

gavv commented on src/Main.cpp:
```diff
@@ -77,6 +77,7 @@ int main(int argc, char** argv) {
 
     opts.add_options("General")
         ("h,help", "Print help message and exit")
+        ("j,json", "Print output in JSON Format")
```
Since we may want to add more formats in future, e.g. CSV, let's replace it with `-f` / `--format` with allowed values `text` and `json`. The short usage would be: `-fjson`, which is still compact.

BitsonFire commented:
Hi @gavv,

Thank You for the comments, they were very helpful. I've made the changes as per the comments. Please let me know of any issues or fixes to be made.

gavv commented on src/JSONFormatter.hpp:
```diff
@@ -0,0 +1,21 @@
+
+        void report_losses(double loss_rate, int sma_window, double avg_loss_rate, double loss_ratio) override;
+
+        JSONFormatter();
+        
+        ~JSONFormatter();
```
It also makes sense to mark copy ctor and assignment operator as deleted. Same for TextFormatter.

gavv commented on src/JSONFormatter.cpp:
```diff
@@ -0,0 +1,43 @@
+void JSONFormatter::report_latency(double sw_hw, double hw, int sma_window, double hw_avg) {
+    if (first_output_) {
+        printf("[\n");
+        printf("  {\"sw+hw\": %7.3f, \"hw\": %7.3f, \"hw_avg%d\": %7.3f}", sw_hw,
```
I suggest to remove limiting float precision from JSON formatter, since it's supposed to be used by programs, not humans, so shorter representation is non needed and even not desired.

Also let's replace `sw+hw` to something like `sw_hw` in JSON to make it a valid identifier, so that one could easily convert it to a class field in dynamic languages like python.

gavv commented on src/LatencyEstimator.hpp:
```diff
@@ -20,11 +22,13 @@ namespace signal_estimator {
 class LatencyEstimator : public IEstimator {
 public:
-    LatencyEstimator(const Config& config);
+    LatencyEstimator(const Config& config, std::unique_ptr<IFormatter>& formatter);
```
Here and in LossEstimator.

Passing unique_ptr by non-constant reference assumes that the callee (ctor) is allowed modify the passed pointer, e.g. reset it to another value.

Since it's not the case, I suggest to replace it with either of these:

* passing unique_ptr by value
* passing unique_ptr by constant reference
* passing raw reference IFormatter

gavv commented:
Could you please also update README? We should update "Command-line options" section, and also add a new section about JSON output (I suggest to add it right after "Measuring losses" section).

gavv commented:
I've tested your branch and it works good, but there is a bug, not related to this PR, which doesn't allow to use JSON properly. I've created a new issue for it: #5.

BitsonFire commented:
I've updated the files based on your suggestions. I've also removed the '%' from the ratio attribute of the JSONFormatter::report_losses. Also please look through the Readme file and let me know if the updated section is sufficient.
Thank you!

gavv commented on src/LatencyEstimator.hpp:
```diff
@@ -82,6 +86,8 @@ class LatencyEstimator : public IEstimator {
     Timestamp input_ts_ {};
     SmaCounter sma_;
+
+    const std::unique_ptr<IFormatter>& format_;
```
Storing a reference or pointer to `unique_ptr<IFormatter>` doesn't provide any extra safety compared to storing raw reference or pointer to `IFormatter`. If we don't store the unique_ptr by value, we don't get the guarantees its intended to provide (retain the ownership).

Such usage works in this case but is confusing: why we use unique_ptr here at all? is it a bug? It's better either to store unique_ptr by value (if we want to take the ownership) or to store a raw reference / pointer to IFormatter (if we don't need the ownership).

Also, sorry, I spoke too soon and my suggestion to use constant reference was actually meaningless. I was thinking about shared_ptr, in this case passing by constant reference would be okay, it'd allow us to make a copy of the shared_ptr and store it in a field. However unique_ptr can't be copied, it can be only moved.

Passing unique_ptr somewhere makes sense only when we want to *transfer* ownership. This can be done either by passing it by value and using std::move on the caller side, or by passing it by lvalue (`&`) or rvalue (`&&`) reference and using std::move in the called function.

The last two options are usually *not* preferred to transfer ownership because they don't make the fact of the transferring obvious. When you read the call, you see that a unique_ptr is passed by reference and thus *may* be modified in the called function, but you don't know whether the it will be actually moved or not, without reading the source code or documentation of the function.

So the good options here actually are:

* store raw reference, pass raw reference (ownership not transferred)
* store unique_ptr by value, pass unique_ptr by value, use std::move outside of ctor (ownership is transferred)

gavv commented on README.md:
````diff
@@ -232,6 +233,34 @@ These numbers may be rough enough.
+]
+```
+
+Note: Here `sw_hw` means `sw+hw` - total software + hardware latency, including ALSA ring buffer. Except `sw_hw`, all the notations are the same as mentioned in the measuring latency and losses section.
````
It's also worth to mention that all time units are in milliseconds.

gavv commented:
LGTM, thanks!

BitsonFire commented:
Wonderful! Thanks for the experience and guidance @gavv. I've learnt a lot in through this PR. :smile:
